# როგორ დავაინსტალიროთ Pandas

In [1]:
!pip install pandas

# Pandas-ს და სხვა საჭირო ბიბლიოთეკების შემოტანა

In [1]:
import pandas as pd

# სხვადასხვა ტიპის ფაილის წაკითხვა

In [2]:
train_df = pd.read_csv('https://docs.google.com/spreadsheets/d/e/2PACX-1vQ-DtJ_iRrgZAA-PU8-ecV6yCGJCrf4nbOcQqtPUKALs7PGm111QKb5el8bxsXeZ370zgDKHeevCcF8/pub?gid=1654560463&single=true&output=csv')

In [4]:
train_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# შემოტანილი ცხრილის დათვალიერება

In [5]:
#სვეტების სახელების სია
train_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [6]:
#ცხრილის განზომილებების ნახვა
train_df.shape

(891, 12)

In [7]:
#პირველი მნიშვნელობა გვიჩვენებს რამდენი ჩანაწერი გვაქვს ცხრილში, ხოლო მეორე მნიშვნელობა აჩვენებს სვეტების რაოდენობას
print("rows:", train_df.shape[0])
print("columns:", train_df.shape[1])

rows: 891
columns: 12


In [8]:
#რა ტიპის სვეტები გვაქვს ცხრილში
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [9]:
#ცხრილის სტატისტიკური მნიშვნელობების დათვალიერება რიცხვით ცვლადებზე
train_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

# მონაცემების სელექცია და ფილტრაცია

## სვეტების მიხედვით სელექცია

In [10]:
columns_to_select = ['Name','Age']

In [11]:
train_df[columns_to_select]

Name   Age
0                              Braund, Mr. Owen Harris  22.0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0
2                               Heikkinen, Miss. Laina  26.0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0
4                             Allen, Mr. William Henry  35.0
..                                                 ...   ...
886                              Montvila, Rev. Juozas  27.0
887                       Graham, Miss. Margaret Edith  19.0
888           Johnston, Miss. Catherine Helen "Carrie"   NaN
889                              Behr, Mr. Karl Howell  26.0
890                                Dooley, Mr. Patrick  32.0

[891 rows x 2 columns]

In [12]:
train_df[['Name','Age']]

Name   Age
0                              Braund, Mr. Owen Harris  22.0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0
2                               Heikkinen, Miss. Laina  26.0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0
4                             Allen, Mr. William Henry  35.0
..                                                 ...   ...
886                              Montvila, Rev. Juozas  27.0
887                       Graham, Miss. Margaret Edith  19.0
888           Johnston, Miss. Catherine Helen "Carrie"   NaN
889                              Behr, Mr. Karl Howell  26.0
890                                Dooley, Mr. Patrick  32.0

[891 rows x 2 columns]

## ჩანაწერების მიხედვით ამორჩევა

In [13]:
start = 2 
end = 5

In [14]:
#საწყისი და საბოლოო ინდექსით(საიდან-სადამდე)
train_df.iloc[start:end]

PassengerId  Survived  Pclass  \
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
2                        Heikkinen, Miss. Laina  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket    Fare Cabin Embarked  
2  STON/O2. 3101282   7.925   NaN        S  
3            113803  53.100  C123        S  
4            373450   8.050   NaN        S

In [15]:
# მხოლოდ საწყისი ინდექსით(საიდან-ბოლო ჩანაწერამდე)
train_df.iloc[start:]

PassengerId  Survived  Pclass  \
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
2                          Heikkinen, Miss. Laina  female  26.0      0      0   
3    Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                        Allen, Mr. William Henry    male  35.0      0      0   
5                                Moran, Mr. James    male   NaN      0      0   
6                         McCarthy, Mr. Timothy J    male  54.0      0      0   
..                                            ...     ...   ...    ...    ...   
886                         Montvila, Rev. Juozas    male  27.0      0      0   
887                  Graham, Miss. Margaret Edith  female  19.0      0      0   
888      Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1      2   
889                         Behr, Mr. Karl Howell    male  26.0      0      0   
890                           Dooley, Mr. Patrick    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  
2    STON/O2. 3101282   7.9250   NaN        S  
3              113803  53.1000  C123        S  
4              373450   8.0500   NaN        S  
5              330877   8.4583   NaN        Q  
6               17463  51.8625   E46        S  
..                ...      ...   ...      ...  
886            211536  13.0000   NaN        S  
887            112053  30.0000   B42        S  
888        W./C. 6607  23.4500   NaN        S  
889            111369  30.0000  C148        C  
890            370376   7.7500   NaN        Q  

[889 rows x 12 columns]

# sadadsa


In [16]:
#მხოლოდ საბოლოო ინდექსით(პირველი ჩანაწერიდან - სადამდე)
train_df[:end]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## სვეტებით და ჩანაწერებით ამორჩევა ერთდროულად

In [17]:
train_df.iloc[start:end, : ]

PassengerId  Survived  Pclass  \
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
2                        Heikkinen, Miss. Laina  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket    Fare Cabin Embarked  
2  STON/O2. 3101282   7.925   NaN        S  
3            113803  53.100  C123        S  
4            373450   8.050   NaN        S

In [18]:
train_df.iloc[start:end, 0:4 ]

PassengerId  Survived  Pclass                                          Name
2            3         1       3                        Heikkinen, Miss. Laina
3            4         1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)
4            5         0       3                      Allen, Mr. William Henry

In [19]:
train_df.iloc[start:end, 0: ]

PassengerId  Survived  Pclass  \
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
2                        Heikkinen, Miss. Laina  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket    Fare Cabin Embarked  
2  STON/O2. 3101282   7.925   NaN        S  
3            113803  53.100  C123        S  
4            373450   8.050   NaN        S

In [20]:
train_df.iloc[start:end, :4 ]

PassengerId  Survived  Pclass                                          Name
2            3         1       3                        Heikkinen, Miss. Laina
3            4         1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)
4            5         0       3                      Allen, Mr. William Henry

## პირობით ფილტრაცია

In [21]:
#შევქმნათ ერთი პირობა
(train_df.Age>44)
#ან
(train_df['Age']>44)

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

### train_df['Age']>44 ან train_df.Age>44 პირობა, აბრუნებს True და False ების სერიას, ჩვენ შეგვიძლია ეს სერია გამოვიყენოთ როგორც ფილტრი და ამოვიღოთ მხოლოდ ის ჩანაწერები სადაც პირობა სრულდება, ანუ სერიაში True მნიშვნელობა გვაქვს

In [22]:
train_df[(train_df["Age"] > 44)]

PassengerId  Survived  Pclass  \
6              7         0       1   
11            12         1       1   
15            16         1       2   
33            34         0       2   
52            53         1       1   
..           ...       ...     ...   
857          858         1       1   
862          863         1       1   
871          872         1       1   
873          874         0       3   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
33                               Wheadon, Mr. Edward H    male  66.0      0   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
..                                                 ...     ...   ...    ...   
857                             Daly, Mr. Peter Denis     male  51.0      0   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch      Ticket     Fare Cabin Embarked  
6        0       17463  51.8625   E46        S  
11       0      113783  26.5500  C103        S  
15       0      248706  16.0000   NaN        S  
33       0  C.A. 24579  10.5000   NaN        S  
52       0    PC 17572  76.7292   D33        C  
..     ...         ...      ...   ...      ...  
857      0      113055  26.5500   E17        S  
862      0       17466  25.9292   D17        S  
871      1       11751  52.5542   D35        S  
873      0      345765   9.0000   NaN        S  
879      1       11767  83.1583   C50        C  

[115 rows x 12 columns]

In [23]:
#ჩვენ შეგვიძლია შევინახოთ ჩვენი პირობა ცვლადში და შემდეგ გამოვიყენოთ
condition = (train_df.Age>44)

In [24]:
train_df[condition]

PassengerId  Survived  Pclass  \
6              7         0       1   
11            12         1       1   
15            16         1       2   
33            34         0       2   
52            53         1       1   
..           ...       ...     ...   
857          858         1       1   
862          863         1       1   
871          872         1       1   
873          874         0       3   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
33                               Wheadon, Mr. Edward H    male  66.0      0   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
..                                                 ...     ...   ...    ...   
857                             Daly, Mr. Peter Denis     male  51.0      0   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch      Ticket     Fare Cabin Embarked  
6        0       17463  51.8625   E46        S  
11       0      113783  26.5500  C103        S  
15       0      248706  16.0000   NaN        S  
33       0  C.A. 24579  10.5000   NaN        S  
52       0    PC 17572  76.7292   D33        C  
..     ...         ...      ...   ...      ...  
857      0      113055  26.5500   E17        S  
862      0       17466  25.9292   D17        S  
871      1       11751  52.5542   D35        S  
873      0      345765   9.0000   NaN        S  
879      1       11767  83.1583   C50        C  

[115 rows x 12 columns]

###  '~' ასრულებს ლოგიკური უარყოფის ფუნქციას, მაგალითად: 
### თუ ჩვენი პირობა არის train_df.Age > 44, 
### ~train_df.Age > 44 იქნება იგივე რაც train_df.Age <= 44

In [25]:
train_df[~condition]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[776 rows x 12 columns]

In [26]:
# პირობა ჩვენ შეგვიძლია შევქმნათ რამდენიმე ქვეპირობისგან, მაგალითად გვჭირდება ამოვიღოთ მხოლოდ 44 წელს ზემოთ მამაკაცები
condition_and = (train_df.Age>44) & (train_df.Sex=='male')
# შესაძლებელია გვჭირდებოდეს ან პირობით გადაბმა, მაგალითად მგზავრები რომლებიც არიან 44 წელს ზემოთ ან არიან მამაკაცები
condition_or = (train_df.Age>44) | (train_df.Sex=='male')

In [27]:
train_df[condition_and]

PassengerId  Survived  Pclass                            Name   Sex  \
6              7         0       1         McCarthy, Mr. Timothy J  male   
33            34         0       2           Wheadon, Mr. Edward H  male   
54            55         0       1  Ostby, Mr. Engelhart Cornelius  male   
62            63         0       1     Harris, Mr. Henry Birkhardt  male   
92            93         0       1     Chaffee, Mr. Herbert Fuller  male   
..           ...       ...     ...                             ...   ...   
771          772         0       3         Jensen, Mr. Niels Peder  male   
789          790         0       1        Guggenheim, Mr. Benjamin  male   
851          852         0       3             Svensson, Mr. Johan  male   
857          858         1       1          Daly, Mr. Peter Denis   male   
873          874         0       3     Vander Cruyssen, Mr. Victor  male   

      Age  SibSp  Parch       Ticket     Fare    Cabin Embarked  
6    54.0      0      0        17463  51.8625      E46        S  
33   66.0      0      0   C.A. 24579  10.5000      NaN        S  
54   65.0      0      1       113509  61.9792      B30        C  
62   45.0      1      0        36973  83.4750      C83        S  
92   46.0      1      0  W.E.P. 5734  61.1750      E31        S  
..    ...    ...    ...          ...      ...      ...      ...  
771  48.0      0      0       350047   7.8542      NaN        S  
789  46.0      0      0     PC 17593  79.2000  B82 B84        C  
851  74.0      0      0       347060   7.7750      NaN        S  
857  51.0      0      0       113055  26.5500      E17        S  
873  47.0      0      0       345765   9.0000      NaN        S  

[79 rows x 12 columns]

In [28]:
train_df[condition_or]

PassengerId  Survived  Pclass                            Name   Sex  \
0              1         0       3         Braund, Mr. Owen Harris  male   
4              5         0       3        Allen, Mr. William Henry  male   
5              6         0       3                Moran, Mr. James  male   
6              7         0       1         McCarthy, Mr. Timothy J  male   
7              8         0       3  Palsson, Master. Gosta Leonard  male   
..           ...       ...     ...                             ...   ...   
883          884         0       2   Banfield, Mr. Frederick James  male   
884          885         0       3          Sutehall, Mr. Henry Jr  male   
886          887         0       2           Montvila, Rev. Juozas  male   
889          890         1       1           Behr, Mr. Karl Howell  male   
890          891         0       3             Dooley, Mr. Patrick  male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
4    35.0      0      0            373450   8.0500   NaN        S  
5     NaN      0      0            330877   8.4583   NaN        Q  
6    54.0      0      0             17463  51.8625   E46        S  
7     2.0      3      1            349909  21.0750   NaN        S  
..    ...    ...    ...               ...      ...   ...      ...  
883  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
886  27.0      0      0            211536  13.0000   NaN        S  
889  26.0      0      0            111369  30.0000  C148        C  
890  32.0      0      0            370376   7.7500   NaN        Q  

[613 rows x 12 columns]

## ფილტრაცია და კონკრეტული სვეტ(ებ)ის ამორჩევა

# დაჯგუფება

## თუ დავაკვირდებით, ტიტანიკზე არსებობდა 3 კლასის ბილეთი, პირველი, მეორე და მესამე. ვნახოთ მათი განაწილება

In [29]:
train_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [30]:
train_df.groupby('Pclass')['Pclass'].count()

Pclass
1    216
2    184
3    491
Name: Pclass, dtype: int64

In [31]:
train_df.Pclass.value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

## სულ გაყიდულა 216 პირველი კლასის ბილეთი, 184 მეორე და 491 მესამე კლასის ბილეთი

## დავაჯგუფოთ კლასის მიხედვით და ვნახოთ რომელი კლასის მგზავრებიდან რამდენი პროცენტი გადარჩა

In [32]:
train_df.groupby(['Pclass'])['Survived'].mean()

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

## როგორც ვხედავთ, პირველი კლასიდან 63%მდე გადარჩა, მეორე კლასის მგზავრების 47% ხოლო მესამედან ყველაზე ცოტა, 24%

## ახლა გავარკვიოთ რომელ სქესის ადამიანებიდან გადარჩნენ უფრო დიდი პროცენტი

In [33]:
train_df.groupby(['Sex'])['Survived'].mean()

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

## გადარჩენილა ქალების 74% ხოლო მამაკაცებიდან დაახლოებით 19%

## ახლა კი გავაერთიანოთ სქესი, ბილეთის კლასი და ორ ჭრილში შევხედოთ ერთდროულად

In [34]:
train_df.groupby(['Pclass','Sex'])['Survived'].mean()

Pclass  Sex   
1       female    0.968085
        male      0.368852
2       female    0.921053
        male      0.157407
3       female    0.500000
        male      0.135447
Name: Survived, dtype: float64

In [35]:
train_df.groupby(['Sex','Pclass'])['Survived'].mean()

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64

# მანიპულაციები სვეტებზე


## შევხედოთ ასაკის ცვლადის განაწილებას

In [36]:
train_df.Age.describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

## მოდით ვნახოთ გადარჩენის პროცენტული რაოდენობა ასაკობრივ ჭრილში

In [37]:
train_df.groupby(['Age'])['Survived'].mean()

Age
0.42     1.0
0.67     1.0
0.75     1.0
0.83     1.0
0.92     1.0
        ... 
70.00    0.0
70.50    0.0
71.00    0.0
74.00    0.0
80.00    1.0
Name: Survived, Length: 88, dtype: float64

## როგორც ვხედავთ, საკმაოდ გაუგებარი და რთულად აღსაქმელი ცხრილი მივიღეთ, სქესის, ან მგზავრობის ბილეთისგან განსხვავებით, ასაკი მერყეობს 0.42-დან 80მდე, რაც საკმაოდ ბევრია. შევქმნათ ახალი ცვლადი, დავყოთ 20 წლიან ინტერვალებად და ისე შევხედოთ.

In [38]:
train_df.Age.value_counts()

24.00    30
22.00    27
18.00    26
19.00    25
30.00    25
         ..
55.50     1
70.50     1
66.00     1
23.50     1
0.42      1
Name: Age, Length: 88, dtype: int64

In [39]:
cut_ages = ['-18', "18-30", '31-50', '50+',]
cut_bins = [0, 18, 30, 50,100]
train_df['Age_bins'] = pd.cut(train_df['Age'], bins=cut_bins, labels = cut_ages)


In [40]:
train_df[['Age','Age_bins']]

Age Age_bins
0    22.0    18-30
1    38.0    31-50
2    26.0    18-30
3    35.0    31-50
4    35.0    31-50
..    ...      ...
886  27.0    18-30
887  19.0    18-30
888   NaN      NaN
889  26.0    18-30
890  32.0    31-50

[891 rows x 2 columns]

In [41]:
train_df.Age_bins.describe()

count       714
unique        4
top       18-30
freq        270
Name: Age_bins, dtype: object

In [42]:
train_df.groupby(['Age_bins'])['Survived'].mean()

Age_bins
-18      0.503597
18-30    0.355556
31-50    0.423237
50+      0.343750
Name: Survived, dtype: float64

In [43]:
train_df.groupby(['Age_bins','Sex'])['Survived'].mean()

Age_bins  Sex   
-18       female    0.676471
          male      0.338028
18-30     female    0.755556
          male      0.155556
31-50     female    0.779070
          male      0.225806
50+       female    0.941176
          male      0.127660
Name: Survived, dtype: float64

In [44]:
train_df.groupby(['Pclass','Sex','Age_bins'])['Survived'].agg(['count','mean'])

count      mean
Pclass Sex    Age_bins                 
1      female -18          11  0.909091
              18-30        24  0.958333
              31-50        37  0.972973
              50+          13  1.000000
       male   -18           5  0.800000
              18-30        21  0.428571
              31-50        49  0.448980
              50+          26  0.192308
2      female -18          14  1.000000
              18-30        30  0.900000
              31-50        27  0.925926
              50+           3  0.666667
       male   -18          15  0.600000
              18-30        37  0.027027
              31-50        35  0.114286
              50+          12  0.083333
3      female -18          43  0.511628
              18-30        36  0.500000
              31-50        22  0.272727
              50+           1  1.000000
       male   -18          51  0.215686
              18-30       122  0.147541
              31-50        71  0.126761
              50+           9  0.000000

### ვიცით რომ ტიტანიკზე 891 მგზავრი იყო, ხოლო აქ 714-ს გვიჩვენებს, რაც ნიშნავს რომ 177 მგზავრის ასაკი არაა შევსებული.

## ჩვენ დავუშვით ერთი შეცდომა, კერძოდ, აქაც არ ვითვალისწინებთ იმ ადამიანებს ვისი ასაკიც არ ვიცით

### ვნახოთ ჩანაწერები სადაც არ ვიცით ასაკი

In [45]:
condition = train_df.Age.isna()

In [46]:
train_df[condition]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked Age_bins  
5      male  NaN      0      0      330877   8.4583   NaN        Q      NaN  
17     male  NaN      0      0      244373  13.0000   NaN        S      NaN  
19   female  NaN      0      0        2649   7.2250   NaN        C      NaN  
26     male  NaN      0      0        2631   7.2250   NaN        C      NaN  
28   female  NaN      0      0      330959   7.8792   NaN        Q      NaN  
..      ...  ...    ...    ...         ...      ...   ...      ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C      NaN  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S      NaN  
868    male  NaN      0      0      345777   9.5000   NaN        S      NaN  
878    male  NaN      0      0      349217   7.8958   NaN        S      NaN  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S      NaN  

[177 rows x 13 columns]

In [47]:
train_df.Age_bins.dtype

CategoricalDtype(categories=['-18', '18-30', '31-50', '50+'], ordered=True)

In [48]:
train_df.Age_bins = train_df.Age_bins.cat.add_categories('Unknown') #დავამატოთ ახალი კატეგორია

In [49]:
train_df.Age_bins.dtype

CategoricalDtype(categories=['-18', '18-30', '31-50', '50+', 'Unknown'], ordered=True)

### შევავსოთ ცარიელი ასაკის ინტერვალები unknown კატეგორიით

In [50]:
train_df.Age_bins.fillna('Unknown', inplace = True)

### და შევამოწმოთ

In [51]:
train_df[condition]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked Age_bins  
5      male  NaN      0      0      330877   8.4583   NaN        Q  Unknown  
17     male  NaN      0      0      244373  13.0000   NaN        S  Unknown  
19   female  NaN      0      0        2649   7.2250   NaN        C  Unknown  
26     male  NaN      0      0        2631   7.2250   NaN        C  Unknown  
28   female  NaN      0      0      330959   7.8792   NaN        Q  Unknown  
..      ...  ...    ...    ...         ...      ...   ...      ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  Unknown  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  Unknown  
868    male  NaN      0      0      345777   9.5000   NaN        S  Unknown  
878    male  NaN      0      0      349217   7.8958   NaN        S  Unknown  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  Unknown  

[177 rows x 13 columns]

In [52]:
train_df.groupby(['Pclass','Sex','Age_bins'])['Survived'].agg(['count','mean'])

count      mean
Pclass Sex    Age_bins                 
1      female -18          11  0.909091
              18-30        24  0.958333
              31-50        37  0.972973
              50+          13  1.000000
              Unknown       9  1.000000
       male   -18           5  0.800000
              18-30        21  0.428571
              31-50        49  0.448980
              50+          26  0.192308
              Unknown      21  0.238095
2      female -18          14  1.000000
              18-30        30  0.900000
              31-50        27  0.925926
              50+           3  0.666667
              Unknown       2  1.000000
       male   -18          15  0.600000
              18-30        37  0.027027
              31-50        35  0.114286
              50+          12  0.083333
              Unknown       9  0.222222
3      female -18          43  0.511628
              18-30        36  0.500000
              31-50        22  0.272727
              50+           1  1.000000
              Unknown      42  0.595238
       male   -18          51  0.215686
              18-30       122  0.147541
              31-50        71  0.126761
              50+           9  0.000000
              Unknown      94  0.095745

## განვიხილოთ ცარიელი ცვლადების შევსების სხვადასხვა მეთოდები, ასაკის ცვლადის მაგალითზე

In [53]:
condition = train_df.Age.isna()
train_df[condition]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked Age_bins  
5      male  NaN      0      0      330877   8.4583   NaN        Q  Unknown  
17     male  NaN      0      0      244373  13.0000   NaN        S  Unknown  
19   female  NaN      0      0        2649   7.2250   NaN        C  Unknown  
26     male  NaN      0      0        2631   7.2250   NaN        C  Unknown  
28   female  NaN      0      0      330959   7.8792   NaN        Q  Unknown  
..      ...  ...    ...    ...         ...      ...   ...      ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  Unknown  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  Unknown  
868    male  NaN      0      0      345777   9.5000   NaN        S  Unknown  
878    male  NaN      0      0      349217   7.8958   NaN        S  Unknown  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  Unknown  

[177 rows x 13 columns]

In [54]:
print('ასაკის საშუალო',train_df.Age.describe()['mean'])
print('ასაკის მედიანა',train_df.Age.describe()['50%'])

ასაკის საშუალო 29.69911764705882
ასაკის მედიანა 28.0


In [55]:
train_df.Age.fillna(train_df.Age.mean()).describe() # ასე შეგვიძლია ცარიელი ასაკები შევავსოთ ასაკის საშუალო მნიშვნელობით

count    891.000000
mean      29.699118
std       13.002015
min        0.420000
25%       22.000000
50%       29.699118
75%       35.000000
max       80.000000
Name: Age, dtype: float64

In [56]:
train_df.Age.fillna(train_df.Age.median()).describe() # ან მედიანით, თუ დავაკვირდებით მედიანით შეცვლამ ჩვენს შემთხვევაში ჩამოწია საშუალო ასაკი

count    891.000000
mean      29.361582
std       13.019697
min        0.420000
25%       22.000000
50%       28.000000
75%       35.000000
max       80.000000
Name: Age, dtype: float64

In [ ]:
ქალების შევავსოთ ქალების საშუალო ასაკით და კაცების კაცების საშუალოთი.

# merge

## playing with types

 https://pandas.pydata.org/docs/getting_started/index.html#getting-started